Add referrer inputs for user during account registration

cpeel opened the pull request:
This started out as a simple request by @lhamilton1 to allow users to specify where they found out about us during account creation. We got there, but not before a bit of refactoring. This PR thus includes:

* Moving logic that interacts with the non_activated_users table into a NonactivatedUser class and corresponding code updates.
* Minor database change to non_activated_users table to better align with the users table.
* Registration styling tweaks.
* Storing the HTTP_REFERER in a cookie for users who are not logged in to make it available in case they end up creating an account. This isn't always set by the browser, but seems like it could be useful for the times we get it.
* *Finally*, adding the requested input field in the registration form.

Right now there is no interface to show the saved data, but we can at least start tracking it for later analysis.

Note, the changes to `addproofer.php` aren't as severe as it seems from the diff. Most of that is from moving `_validate_fields()` to the end of the file.

wf49670 commented on pinc/dpsession.inc:
```diff
@@ -71,6 +73,17 @@ function dpsession_resume()
         _update_user_activity_time( FALSE );
     }
+
+    // stash the http_referer from the user's very first page load of the site
```
Elsewhere in this PR we use "referrer" rather than "referer". Is there some reason for this inconsistency?

cpeel replied:
Funny you should ask: https://en.wikipedia.org/wiki/HTTP_referer

cpeel replied:
The `http_referer` cookie is being set here in `pinc/dpsession.inc`. The contents of that cookie are being saved in the `non_activated_users.http_referrer` column in `accounts/addproofer.php` and, upon activation, moved into the `users` table in `accounts/activate.php`.

wf49670 replied:
Got it. Thanks. Consistency of spelling would be nice... Maybe our internal field should be http_referer, too?

cpeel replied:
I thought about that, but then decided it would be very weird to see right next to the other `referrer` field. I opted to push the weirdness at the point of consumption (when we get the HTTP_REFERER from Apache) rather than propagate that throughout the code and database. I'm not sure there is a good decision - I can easily see both sides of it.

wf49670 commented:
Other than comments above, this looks good to me.

cpeel commented:
FWIW, I don't feel that we can roll this out to PROD until we have our Privacy policy updated. That's in front of the DPF board right now. I'd prefer to hold off merging this until that's been approved. (FYI @lhamilton1 )

jmdyck commented on accounts/addproofer.php:
```diff
@@ -77,172 +246,29 @@ function _validate_fields($real_name, $username, $userpass, $userpass2, $email,
-        $error = $func($real_name, $username, $userpass, $userpass2, $email, $email2, $email_updates);
-        if(!empty($error))
-            break;
+        return _("Please tell us how you heard from us.");
```
"from"? Presumably they haven't heard anything *from* us yet. Change to "about"?

cpeel commented:
> Case #2 seems okay at first. It gives me the expectation that it will either:
> 
> return a 'fully-loaded' record of an existing non-activated user whose username is $username, or else
> throw an exception.
> But in fact there's a third possibility: if $username is falsy, it will create an empty NonactivatedUser instance, like case #1. This possibility doesn't seem useful to me, and might be a source of bugs. (It surprised me when I stumbled over it with the User class recently.)

I agree that the case where $username is falsy is a bug and should be fixed both here and in the User class.

> Case #3 has a few things that bother me:
> 
> It takes two statements to accomplish what should be a single expression.
> It requires that load() be a public method, which could be a source of bugs. (E.g., code might call it twice on the same instance, where the second call would overwrite the first.)
> load() is more general than necessary (accepting any field-name as the first arg), which might open it to abuse. (E.g., load("email", $address) would often work, until it doesn't.)

What you see as a problem I see as a design feature. The load() function handles the case for a field with non-unique values and throws an exception in that case. That lets the user of this function use it and get known results back. Your second concern about calling twice on the same instance is valid and we could resolve it by setting a static value if the load() function has been called already and throwing an exception if it was called again on the same object, preventing it from being called twice.

I'm not really a fan of static class functions that return instances of themselves. That seems very un-OO-like and unintuitive to me. Using the constructor to load an instance of an object is a fairly-well established pattern in the DP code (Settings, Project, etc).

cpeel commented on SETUP/tests/NonactivatedUserTest.php:
```diff
@@ -0,0 +1,93 @@
+    public function testLoadRegistration()
+    {
+        $this->createRecord($this->TEST_USERNAME);
+        $register = new NonactivatedUser($this->TEST_USERNAME);
```
But they're the same thing - both return a NonactivatedUser record. I'll rename the function to be `createNonactivatedUser()` and consistently use `$user` for the holding variable.

jmdyck replied:
Those both look like improvements to me. I think the introduction of `$existing_user` helps address what I was complaining about.

cpeel commented on SETUP/tests/NonactivatedUserTest.php:
```diff
@@ -0,0 +1,93 @@
+            WHERE username = '$this->TEST_USERNAME'
+        ";
+        $result = mysqli_query(DPDatabase::get_connection(), $sql);
+    }
```
Good catch, fixed.

jmdyck replied:
Yup.

cpeel commented on accounts/activate.php:
```diff
@@ -117,20 +102,21 @@ function validate_userID($param_name, $value)
-mysqli_query(DPDatabase::get_connection(), "
-    DELETE FROM non_activated_users WHERE id='$ID'
-");
+$user->delete();
```
Good find. It's easy enough to move the non_activate_user deletion after we've created the user record.

jmdyck replied:
So now there's a small window in which the username exists in *both* tables. But I don't *think* that leads to a problem, even if the server crashes within that window.

cpeel commented on pinc/dpsession.inc:
```diff
@@ -71,6 +73,17 @@ function dpsession_resume()
+            setcookie('http_referer', $_SERVER["HTTP_REFERER"]);
+        else
+            setcookie('http_referer', '');
+    }
```
See other PR on why I don't like using `@`.

cpeel commented on pinc/NonactivatedUser.inc:
```diff
@@ -0,0 +1,188 @@
+        mysqli_free_result($result);
+    }
+
+    public function save()
```
I think you are technically correct, however the odds of two non-bots attempting to register the same username at the same time is ridiculously small.

jmdyck replied:
Yeah, probably. We do occasionally receive duplicated requests (from non-bots), which is a more likely scenario for hitting this race condition -- but since the requests would be identical, you'd get a successful + correct registration regardless?

Anyhow, I think the "ON DUPLICATE KEY" deserves a comment in the code: someone could reasonably look at that and think "wait, if the key already exists in the table, isn't that a bug?" And once you've explained why it's there, you could maybe mention that it also allows a ridiculously unlikely race condition.

cpeel replied:
I've updated the save unit test to detect this, confirmed it was broken, updated the code, and confirmed it is fixed.

jmdyck replied:
> The primary key for that row is the code-generated id (userid + 32-character hex string)

Hm? In the `non_activated_users` table, the primary key is the `username` column, not the `id` column. Or am I looking at the wrong thing?

cpeel replied:
You are correct, I don't know what I was looking at.

jmdyck replied:
(So then what did the unit test detect?)

cpeel replied:
The unit test detected that we were always creating a new ID upon every save and that we no longer do that now.

cpeel replied:
Updated function comment with mention about race condition since the primary key is the username.

cpeel commented:
I believe I've addressed most of the PR feedback, with the exception of the philosophical differences in how data should be loaded into a class object upon construction. I've added my updates as additional commits, two of which will be rebased as fixups into prior commits but I've left them separate for reviewability.

jmdyck commented:
Okay, I think I've reviewed all the feedback. (Mind you, I'm finding it difficult to navigate all the commits + comments of this PR.)

cpeel commented on accounts/addproofer.php:
```diff
@@ -28,19 +36,8 @@
 $error = "";
 
-$password = isset($_POST['password'])? $_POST['password']: '';
-if ($password=="proofer") {
```
Probably, but I'm not going to break it apart now.

cpeel commented:
@jmdyck - I've taken a compromise position on the `load()` method. I've made the function private, made the constructor accept an optional username, and added a `load_by_id()` function to load by ID. The User class will receive similar feedback in a separate PR. I think this addresses the core of your concerns without making all object access go through a static class.

There are several commits that need to be merged / removed which I will do before this gets merged. I left them separate to show the progress.

jmdyck commented:
>  I think this addresses the core of your concerns

I think so too. (It fixes all the things that bothered me about "Case #3" above.)

> without making all object access go through a static class.

Do you mean "without making all object construction go though a static method"?

cpeel commented:
> > without making all object access go through a static class.
> 
> Do you mean "without making all object construction go though a static method"?

Well, no. I meant "without making all object construction that also populates the object go through a static method".

jmdyck commented:
Okay.

cpeel commented:
Last commit/rebase was me adding the `SETUP/upgrade/SETUP/upgrade/12/20180514_add_referrer_columns.php` that is a part of this and had been sitting in my sandbox for the past several months <sigh>.
